src/interchange/src/avro/schema.rs

```diff
@@ -7,13 +7,42 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
+//! Conversion from Avro schemas to Materialize `RelationDesc`s.
+//!
+//! A few notes for posterity on how this conversion happens are in order.
+//!
+//! If the schema is an Avro record, we flatten it to its fields, which become the columns
+//! of the relation.
+//!
+//! Each individual field is then converted to its SQL equivalent. For most types, this
+//! conversion is the obvious one. The only non-trivial counterexample is Avro unions.
+//!
+//! Since Avro types are not nullable by default, the typical way normal (i.e., nullable)
+//! SQL fields are represented in Avro is by a union of the underlying type with the
+//! singleton type { Null }; in Avro schema notation, this is `["null", "TheType"]`.
+//! We shall call union types following this pattern _Nullability-Pattern Unions_.
+//! We shall call all other union types (e.g. `["MyType1", "MyType2"]` or `["null", "MyType1", "MyType2"]`) _Essential Unions_.
+//! Since there is an obvious way to represent Nullability-Pattern Unions, but not Essential Unions, in the SQL type system,
+//! we must handle Essential Unions with a bit of a hack (at least until Materialize supports union or sum types, which may be never).
+//!
+//! When an Essential Union appears as one of the fields of a record, we expand
+//! it to _n_ columns in SQL, where _n_ is the number of non-null variants in the union. These
+//! columns will be given names created by pasting their index at the end of the overall name
+//! of the field. For example, if an Essential Union in a field named `"Foo"` has schema `[int, bool]`, it will expand to the columns `"Foo1": bool, "Foo2": int`.
```

Comment:
```suggestion
//! When an Essential Union appears as one of the fields of a record, we expand
//! it to _n_ columns in SQL, where _n_ is the number of non-null variants in the union. These
//! columns will be given names created by pasting their index at the end of the overall name
//! of the field. For example, if an Essential Union in a field named `"Foo"` has schema `[int, bool]`, it will expand to the columns `"Foo1": bool, "Foo2": int`. There is an implicit constraint upheld be the source pipeline that only one such column will be non-`null` at a time
```